Realm/RLMValue.mm

```diff
@@ -60,65 +76,177 @@ - (RLMPropertyType)rlm_valueType {
     }
 }
+#pragma clang diagnostic pop
+
+- (RLMAnyValueType)rlm_anyValueType {
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
+    return (RLMAnyValueType)self.rlm_valueType;
+#pragma clang diagnostic pop
+}
 
 @end
 
 #pragma mark NSNull
 
 @implementation NSNull (RLMValue)
 
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
 - (RLMPropertyType)rlm_valueType {
     return RLMPropertyTypeAny;
 }
+#pragma clang diagnostic pop
+
+- (RLMAnyValueType)rlm_anyValueType {
+    return RLMAnyValueTypeAny;
+}
 
 @end
 
 #pragma mark NSString
 
 @implementation NSString (RLMValue)
 
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
 - (RLMPropertyType)rlm_valueType {
     return RLMPropertyTypeString;
 }
+#pragma clang diagnostic pop
+
+- (RLMAnyValueType)rlm_anyValueType {
+    return RLMAnyValueTypeString;
+}
 
 @end
 
 #pragma mark NSUUID
 
 @implementation NSUUID (RLMValue)
 
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
 - (RLMPropertyType)rlm_valueType {
     return RLMPropertyTypeUUID;
 }
+#pragma clang diagnostic pop
+
+- (RLMAnyValueType)rlm_anyValueType {
+    return RLMAnyValueTypeUUID;
+}
 
 @end
 
 #pragma mark RLMDecimal128
 
 @implementation RLMDecimal128 (RLMValue)
 
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
 - (RLMPropertyType)rlm_valueType {
     return RLMPropertyTypeDecimal128;
 }
+#pragma clang diagnostic pop
+
+- (RLMAnyValueType)rlm_anyValueType {
+    return RLMAnyValueTypeDecimal128;
+}
 
 @end
 
 #pragma mark RLMObjectBase
 
 @implementation RLMObjectBase (RLMValue)
 
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
 - (RLMPropertyType)rlm_valueType {
     return RLMPropertyTypeObject;
 }
+#pragma clang diagnostic pop
+
+- (RLMAnyValueType)rlm_anyValueType {
+    return RLMAnyValueTypeObject;
+}
 
 @end
 
 #pragma mark RLMObjectId
 
 @implementation RLMObjectId (RLMValue)
 
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
 - (RLMPropertyType)rlm_valueType {
     return RLMPropertyTypeObjectId;
 }
+#pragma clang diagnostic pop
+
+- (RLMAnyValueType)rlm_anyValueType {
+    return RLMAnyValueTypeObjectId;
+}
+
+@end
+
+#pragma mark Dictionary
+
+@implementation NSDictionary (RLMValue)
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
+- (RLMPropertyType)rlm_valueType {
+    REALM_UNREACHABLE();
+}
+#pragma clang diagnostic pop
+
+- (RLMAnyValueType)rlm_anyValueType {
+    return RLMAnyValueTypeDictionary;
+}
+
+@end
+
+@implementation RLMDictionary (RLMValue)
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
+- (RLMPropertyType)rlm_valueType {
+    REALM_UNREACHABLE();
```

Comment:
This is not unreachable. `rlm_valueType` is part of the public API.

Reply:
@tgoyne This is a very particular case because we are deprecating the declaration but also changing the type to be used. You are right because this is part of the public, the user can ask for `rlm_valueType `, but because we don't want to add any new types to property type we are in some sort of limbo here.
Solutions
We can return RLMPropertyTypeAny like we do for Null, and this is actually true because this indeed a colection of mixed?

Reply:
Returning RLMPropertyTypeAny seems like the best option here.

REALM_UNREACHABLE() marks a place that should be logically impossible to hit and if we get there it always indicates a bug in our code.